src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -471,15 +480,72 @@ public ManagedObjectWrapperReleaser(ManagedObjectWrapper* wrapper)
         internal unsafe class NativeObjectWrapper
         {
             internal IntPtr _externalComObject;
+            private IntPtr _inner;
             private ComWrappers _comWrappers;
             internal GCHandle _proxyHandle;
+            internal GCHandle _proxyHandleTrackingResurrection;
+            internal GCHandle _nativeObjectWrapperWeakHandle;
+
+            private readonly bool _aggregatedManagedObjectWrapper;
+            internal readonly bool _fromTrackerRuntime;
+            private IntPtr _trackerObject;
+            private readonly bool _releaseTrackerObject;
+            private int _trackerObjectDisconnected; // Atomic boolean, so using int.
+
+            internal readonly IntPtr _contextToken;
 
-            public NativeObjectWrapper(IntPtr externalComObject, ComWrappers comWrappers, object comProxy)
+            public IntPtr TrackerObject => (_trackerObject == IntPtr.Zero || _trackerObjectDisconnected == 1) ? IntPtr.Zero : _trackerObject;
+
+            public NativeObjectWrapper(IntPtr externalComObject, IntPtr inner, ComWrappers comWrappers, object comProxy, CreateObjectFlags flags)
             {
                 _externalComObject = externalComObject;
+                _inner = inner;
                 _comWrappers = comWrappers;
-                Marshal.AddRef(externalComObject);
                 _proxyHandle = GCHandle.Alloc(comProxy, GCHandleType.Weak);
+
+                // We have a separate handle tracking resurrection as we want to make sure
+                // we clean up the NativeObjectWrapper only after the RCW has been finalized
+                // due to it can access the native object in the finalizer. At the same time,
+                // we want other callers which are using _proxyHandle such as the RCW cache to
+                // see the object as not alive once it is eligible for finalization.
+                _proxyHandleTrackingResurrection = GCHandle.Alloc(comProxy, GCHandleType.WeakTrackResurrection);
+
+                _nativeObjectWrapperWeakHandle = GCHandle.Alloc(this, GCHandleType.Weak);
+
+                if (flags.HasFlag(CreateObjectFlags.TrackerObject))
+                {
+                    if (Marshal.QueryInterface(externalComObject, in IID_IReferenceTracker, out IntPtr trackerObject) == 0)
```

Comment:
Should these `0`-s for QI calls use `HResults.S_OK` for readability?

Reply:
Addressed.